src/compiler/emitter.ts

```diff
@@ -1639,10 +1639,12 @@ module ts {
             }
 
             if (root) {
+                currentSourceFile = root;
                 emit(root);

```

Comment:
The problem with the previous form was that setting current would happen far down the line.  i.e. only once we actually got around to calling 'emitSourceFile'.  That meant that between the call to 'emit' and the eventual call to emitSourceFile, you had a time when there was no 'currentSourceFile', which could break anyone who tried to use it for any purpose.

This approach ensures that currentSourceFile is set hte moment we start processing an actual source file.

The alternative is to move this check/set into 'emit'.  However, that means that ever call to emit need to immediately check what type of node it is emitting, and then set currentSourceFile if it's a sourceFile.  Given how often emit is called, it seems better to just set the sourceFile the moment we know we're emitting a file, and hten have it correctly set for all code that runs from that point on.